### ISSUE TITLE
Question in normalization step in prepare MFCC step. 

### ISSUE THREAD

**Issue by RobinROAR** [philipperemy/deep-speaker#63]
Hi Philip, 

I have a two questions about the way your prepare the MFCC feature (audio.py). 

1.  In the following code, you normalized the raw Fbank features (num_frames, nfit ). 

```
 filter_banks, energies = fbank(signal, samplerate=sample_rate, nfilt=NUM_FBANKS)
 frames_features = normalize_frames(filter_banks)
...
def normalize_frames(m, epsilon=1e-12):
    return [(v - np.mean(v)) / max(np.std(v), epsilon) for v in m]
```
But I think your code works in normalizing the 26D Fbank features within just each frame. It is actually the instance normalization but not the commonly-used batch normalization. If we want to normalize data at the batch (or the training data) level, I believe we should do like 

` return [ (v - np.mean(m, axis=0)) / np.std(m, axis=0) for v in m]]` or 

```
from sklearn.preprocessing import StandardScaler
s1 =StandardScaler()
return s1.fit_transform(m)
```

Could you please explain why you normalize data in this way?

2. Could you explain why the function name is read_mfcc() but you actually choose to used the Fbank features which lack the DCT from MFCC features.

Thanks you!

**Comment by philipperemy** [comment]
Hi Robin,

Thank you for your question.

1. Here is the paragraph where the authors mention about the normalization:

![image](https://user-images.githubusercontent.com/4516927/90086742-18ba2600-dd56-11ea-9774-7897990727ec.png)

From: https://arxiv.org/pdf/1705.02304.pdf

It's not crystal clear how they did so I had to make a guess. I used instance normalization as you pointed out and did not normalize with batch normalization. The 26D FBanks get normalized within a single frame for a given speaker. I'm not a big fan of BN because it's batch dependent.

2. Regarding the naming I agree with you. It should be more `read_fbank()` than `read_mfcc()`.

PS: I'm not an expert in audio analysis.

**Comment by RobinROAR** [comment]
Hi Philip, Thanks for your quick reply! 

I'm not a audio expert as well, but I doubt the IN(Instance Normalization) in each frame because I feel it would remove the difference between audio frames like [0,0,0,0,0..] and [100,100,100,100,100]. Besides, to my knowledge in computer vision field, the IN is usually used in style transfer, GAN-based image synthesize or image-to-image translation tasks. When handling classifications, the BN is mostly applied. 

To verify my thoughts, I make a minor revise on your code(use 64D MFCC and do BN) and retrain the model. The test script result is  SAME SPEAKER [0.73709786]  DIFF SPEAKER [-0.00997655]  (yours: SAME SPEAKER [0.8112024]   DIFF SPEAKER [0.02534033]) .  No significant improvement...

I find a similar discuss in https://www.kaggle.com/c/freesound-audio-tagging/discussion/54082
I'm also trying to contact the original paper authors to find the truth.

Best,

-Robin

**Comment by philipperemy** [comment]
@RobinROAR thank you for making those experiments. It's very insightful. I would expect it to work better too but seems like in practice it yields similar results. I tried to contact the original paper authors but it has been unsuccessful. It's a paper by Baidu and they probably don't want to release too much information. I had to "guess" a lot of small details when implementing the paper. Anyway, let me know if they reply!
